File: web/backend/api/v2/finance.py
"""Finance API — собственный учёт финансов инфраструктуры (P&L).

Заменяет прокси к панельному infra-billing: категории, провайдеры, записи
расходов/доходов с валютами и циклами, платежи с фиксацией курса, курсы
валют, сводка P&L и импорт данных из панели.
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field, field_validator

from web.backend.api.deps import AdminUser, require_permission
from shared.database import db_service
from shared.config_service import config_service
from shared.db.finance import BILLING_CYCLES, ITEM_KINDS

logger = logging.getLogger(__name__)
router = APIRouter()


def _base_currency() -> str:
    return str(config_service.get("finance_base_currency", "RUB") or "RUB").upper()


async def _base_rate() -> float:
    """Курс базовой валюты к RUB (RUB -> 1.0)."""
    base = _base_currency()
    if base == "RUB":
        return 1.0
    for r in await db_service.get_finance_rates():
        if r["currency"] == base and r["rate_rub"]:
            return float(r["rate_rub"])
    return 1.0


# ── Pydantic модели ──────────────────────────────────────────────


class CategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    kind: str = "expense"
    color: Optional[str] = None
    icon: Optional[str] = None

    @field_validator("kind")
    @classmethod
    def _kind(cls, v):
        if v not in ITEM_KINDS:
            raise ValueError("kind must be expense|income")
        return v


class CategoryUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    color: Optional[str] = None
    icon: Optional[str] = None
    sort_order: Optional[int] = None


class ProviderCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    url: Optional[str] = None
    favicon_url: Optional[str] = None
    notes: Optional[str] = None


class ProviderUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    url: Optional[str] = None
    favicon_url: Optional[str] = None
    notes: Optional[str] = None


class ItemCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    kind: str = "expense"
    category_id: Optional[int] = None
    provider_id: Optional[int] = None
    node_uuid: Optional[str] = None
    currency: str = Field(default="RUB", min_length=3, max_length=8)
    amount: float = Field(default=0, ge=0)
    billing_cycle: str = "monthly"
    cycle_days: Optional[int] = Field(default=None, ge=1, le=3650)
    next_due_at: Optional[str] = None
    url: Optional[str] = None
    notes: Optional[str] = None

    @field_validator("kind")
    @classmethod
    def _kind(cls, v):
        if v not in ITEM_KINDS:
            raise ValueError("kind must be expense|income")
        return v

    @field_validator("billing_cycle")
    @classmethod
    def _cycle(cls, v):
        if v not in BILLING_CYCLES:
            raise ValueError(f"billing_cycle must be one of {BILLING_CYCLES}")
        return v


class ItemUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    kind: Optional[str] = None
    category_id: Optional[int] = None
    provider_id: Optional[int] = None
    node_uuid: Optional[str] = None
    currency: Optional[str] = Field(default=None, min_length=3, max_length=8)
    amount: Optional[float] = Field(default=None, ge=0)
    billing_cycle: Optional[str] = None
    cycle_days: Optional[int] = Field(default=None, ge=1, le=3650)
    next_due_at: Optional[str] = None
    url: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[str] = None

    @field_validator("kind")
    @classmethod
    def _kind(cls, v):
        if v is not None and v not in ITEM_KINDS:
            raise ValueError("kind must be expense|income")
        return v

    @field_validator("billing_cycle")
    @classmethod
    def _cycle(cls, v):
        if v is not None and v not in BILLING_CYCLES:
            raise ValueError(f"billing_cycle must be one of {BILLING_CYCLES}")
        return v

    @field_validator("status")
    @classmethod
    def _status(cls, v):
        if v is not None and v not in ("active", "archived"):
            raise ValueError("status must be active|archived")
        return v


class MarkPaid(BaseModel):
    amount: Optional[float] = Field(default=None, ge=0)
    paid_at: Optional[str] = None
    comment: Optional[str] = None


class PaymentCreate(BaseModel):
    item_id: Optional[int] = None
    item_name: Optional[str] = Field(default=None, max_length=200)
    kind: str = "expense"
    paid_at: str
    amount: float = Field(gt=0)
    currency: str = Field(default="RUB", min_length=3, max_length=8)
    comment: Optional[str] = None

    @field_validator("kind")
    @classmethod
    def _kind(cls, v):
        if v not in ITEM_KINDS:
            raise ValueError("kind must be expense|income")
        return v


class RateUpdate(BaseModel):
    rate_rub: float = Field(gt=0)
    is_manual: bool = True


# ── Summary / Upcoming ───────────────────────────────────────────


@router.get("/summary")
async def finance_summary(
    months: int = Query(6, ge=1, le=36),
    admin: AdminUser = Depends(require_permission("finance", "view")),
):
    summary = await db_service.finance_summary(months=months)
    base = _base_currency()
    base_rate = await _base_rate()

    def conv(v_rub: float) -> float:
        return round(v_rub / base_rate, 2)

    return {
        "base_currency": base,
        "monthly": [
            {**m, "expense": conv(m["expense_rub"]), "income": conv(m["income_rub"]),
             "net": conv(m["net_rub"])}
            for m in summary["monthly"]
        ],
        "by_category": [
            {**c, "monthly": conv(c["monthly_rub"])} for c in summary["by_category"]
        ],
        "by_currency": summary["by_currency"],
        "recurring": {
            "expense": conv(summary["recurring"].get("expense_rub", 0)),
            "income": conv(summary["recurring"].get("income_rub", 0)),
            "net": conv(summary["recurring"].get("net_rub", 0)),
        },
    }


@router.get("/upcoming")
async def finance_upcoming(
    days: int = Query(30, ge=1, le=365),
    admin: AdminUser = Depends(require_permission("finance", "view")),
):
    items = await db_service.upcoming_finance_payments(days=days)
    return {"items": items, "total": len(items)}


# ── Categories ───────────────────────────────────────────────────


@router.get("/categories")
async def list_categories(admin: AdminUser = Depends(require_permission("finance", "view"))):
    return {"items": await db_service.list_finance_categories()}


@router.post("/categories")
async def create_category(
    data: CategoryCreate,
    admin: AdminUser = Depends(require_permission("finance", "create")),
):
    created = await db_service.create_finance_category(data.name, data.kind, data.color, data.icon)
    if not created:
        raise HTTPException(status_code=409, detail="Category already exists")
    return created


@router.patch("/categories/{category_id}")
async def update_category(
    category_id: int,
    data: CategoryUpdate,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    ok = await db_service.update_finance_category(
        category_id, **data.model_dump(exclude_unset=True, exclude_none=True),
    )
    if not ok:
        raise HTTPException(status_code=404, detail="Category not found")
    return {"status": "ok"}


@router.delete("/categories/{category_id}")
async def delete_category(
    category_id: int,
    admin: AdminUser = Depends(require_permission("finance", "delete")),
):
    ok = await db_service.delete_finance_category(category_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Category not found or is system")
    return {"status": "ok"}


# ── Providers ────────────────────────────────────────────────────


@router.get("/providers")
async def list_providers(admin: AdminUser = Depends(require_permission("finance", "view"))):
    return {"items": await db_service.list_finance_providers()}


@router.post("/providers")
async def create_provider(
    data: ProviderCreate,
    admin: AdminUser = Depends(require_permission("finance", "create")),
):
    created = await db_service.create_finance_provider(
        data.name, data.url, data.favicon_url, data.notes,
    )
    if not created:
        raise HTTPException(status_code=500, detail="Failed to create provider")
    return created


@router.patch("/providers/{provider_id}")
async def update_provider(
    provider_id: int,
    data: ProviderUpdate,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    ok = await db_service.update_finance_provider(
        provider_id, **data.model_dump(exclude_unset=True, exclude_none=True),
    )
    if not ok:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"status": "ok"}


@router.delete("/providers/{provider_id}")
async def delete_provider(
    provider_id: int,
    admin: AdminUser = Depends(require_permission("finance", "delete")),
):
    ok = await db_service.delete_finance_provider(provider_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"status": "ok"}


# ── Items ────────────────────────────────────────────────────────


@router.get("/items")
async def list_items(
    kind: Optional[str] = Query(None),
    status: Optional[str] = Query("active"),
    category_id: Optional[int] = Query(None),
    currency: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    admin: AdminUser = Depends(require_permission("finance", "view")),
):
    items = await db_service.list_finance_items(
        kind=kind, status=status or None, category_id=category_id,
        currency=currency, search=search,
    )
    return {"items": items, "total": len(items)}


@router.post("/items")
async def create_item(
    data: ItemCreate,
    admin: AdminUser = Depends(require_permission("finance", "create")),
):
    item = await db_service.create_finance_item(**data.model_dump())
    if not item:
        raise HTTPException(status_code=500, detail="Failed to create item")
    return item


@router.patch("/items/{item_id}")
async def update_item(
    item_id: int,
    data: ItemUpdate,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    item = await db_service.update_finance_item(item_id, **data.model_dump(exclude_unset=True))
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.delete("/items/{item_id}")
async def delete_item(
    item_id: int,
    admin: AdminUser = Depends(require_permission("finance", "delete")),
):
    ok = await db_service.delete_finance_item(item_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Item not found")
    return {"status": "ok"}


@router.post("/items/{item_id}/paid")
async def mark_paid(
    item_id: int,
    data: MarkPaid,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    item = await db_service.mark_finance_item_paid(
        item_id, amount=data.amount, paid_at=data.paid_at, comment=data.comment,
    )
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.post("/items/{item_id}/skip")
async def skip_cycle(
    item_id: int,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    item = await db_service.skip_finance_item_cycle(item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


# ── Payments ─────────────────────────────────────────────────────


@router.get("/payments")
async def list_payments(
    item_id: Optional[int] = Query(None),
    since: Optional[str] = Query(None),
    until: Optional[str] = Query(None),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    admin: AdminUser = Depends(require_permission("finance", "view")),
):
    items = await db_service.list_finance_payments(
        item_id=item_id, since=since, until=until, limit=limit, offset=offset,
    )
    return {"items": items, "total": len(items)}


@router.post("/payments")
async def create_payment(
    data: PaymentCreate,
    admin: AdminUser = Depends(require_permission("finance", "create")),
):
    """Произвольный платёж/доход без привязки к записи (или с привязкой без сдвига цикла)."""
    from shared.db_schema import FINANCE_PAYMENTS_TABLE, FINANCE_RATES_TABLE
    from datetime import date as _date

    item_name = data.item_name
    if data.item_id and not item_name:
        item = await db_service.get_finance_item(data.item_id)
        item_name = (item or {}).get("name")
    if not item_name:
        item_name = "Без названия"

    async with db_service.acquire() as conn:
        row = await conn.fetchrow(
            f"""INSERT INTO {FINANCE_PAYMENTS_TABLE}
                (item_id, item_name, kind, paid_at, amount, currency, rate_rub, comment, source)
                VALUES ($1, $2, $3, $4::date, $5, $6,
                        (SELECT rate_rub FROM {FINANCE_RATES_TABLE} WHERE currency = $6),
                        $7, 'manual')
                RETURNING id""",
            data.item_id, item_name, data.kind, _date.fromisoformat(data.paid_at),
            round(data.amount, 2), data.currency.upper(), data.comment,
        )
    return {"id": row["id"], "status": "ok"}


@router.delete("/payments/{payment_id}")
async def delete_payment(
    payment_id: int,
    admin: AdminUser = Depends(require_permission("finance", "delete")),
):
    ok = await db_service.delete_finance_payment(payment_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Payment not found")
    return {"status": "ok"}


# ── Rates ────────────────────────────────────────────────────────


@router.get("/rates")
async def list_rates(admin: AdminUser = Depends(require_permission("finance", "view"))):
    return {"items": await db_service.get_finance_rates(), "base_currency": _base_currency()}


@router.put("/rates/{currency}")
async def set_rate(
    currency: str,
    data: RateUpdate,
    admin: AdminUser = Depends(require_permission("finance", "edit")),
):
    await db_service.upsert_finance_rate(currency, data.rate_rub, is_manual=data.is_manual)
    return {"status": "ok"}


@router.post("/rates/refresh")
async def refresh_rates(admin: AdminUser = Depends(require_permission("finance", "edit"))):
    from web.backend.core.finance.rates import update_rates
    updated = await update_rates()
    return {"updated": updated, "items": await db_service.get_finance_rates()}


# ── Import from panel ────────────────────────────────────────────


@router.post("/import-panel")
async def import_panel(admin: AdminUser = Depends(require_permission("finance", "create"))):
    """Одноразовый импорт провайдеров/биллинг-нод/истории из панельного infra-billing."""
    from web.backend.core.finance.importer import import_from_panel
    return await import_from_panel()

File: web/frontend/src/api/finance.ts
import client from './client'

export interface FinanceCategory {
  id: number
  name: string
  kind: 'expense' | 'income'
  color: string | null
  icon: string | null
  is_system: boolean
  sort_order: number
}

export interface FinanceProvider {
  id: number
  name: string
  url: string | null
  favicon_url: string | null
  notes: string | null
  items_count?: number
}

export interface FinanceItem {
  id: number
  kind: 'expense' | 'income'
  name: string
  category_id: number | null
  category_name: string | null
  category_color: string | null
  category_icon: string | null
  provider_id: number | null
  provider_name: string | null
  node_uuid: string | null
  node_name: string | null
  currency: string
  amount: number
  billing_cycle: 'monthly' | 'yearly' | 'once' | 'days'
  cycle_days: number | null
  next_due_at: string | null
  url: string | null
  notes: string | null
  status: 'active' | 'archived'
  monthly_equivalent: number
  days_left?: number
  is_overdue?: boolean
}

export interface FinancePayment {
  id: number
  item_id: number | null
  item_name: string
  kind: 'expense' | 'income'
  paid_at: string
  amount: number
  currency: string
  rate_rub: number | null
  amount_rub: number
  comment: string | null
  source: string
}

export interface FinanceRate {
  currency: string
  rate_rub: number
  is_manual: boolean
  updated_at: string
}

export interface FinanceSummary {
  base_currency: string
  monthly: { month: string; expense: number; income: number; net: number }[]
  by_category: { category: string; color: string; monthly: number }[]
  by_currency: { currency: string; expense_monthly: number; income_monthly: number }[]
  recurring: { expense: number; income: number; net: number }
  this_month?: {
    month: string | null
    expense: number
    income: number
    net: number
    expense_actual: number
    income_actual: number
    expense_upcoming: number
    income_upcoming: number
  }
}

export interface ItemPayload {
  name: string
  kind: string
  category_id?: number | null
  provider_id?: number | null
  currency: string
  amount: number
  billing_cycle: string
  cycle_days?: number | null
  next_due_at?: string | null
  url?: string | null
  notes?: string | null
  status?: string
}

export interface AdapterField {
  name: string
  label: string
  type: 'text' | 'password' | 'url'
  required: boolean
  placeholder: string | null
  help: string | null
}

export interface HosterAdapter {
  slug: string
  title: string
  description: string
  needs_base_url: boolean
  fields: AdapterField[]
}

export interface FinanceService {
  name: string
  status: string | null
  price: number | null
  currency: string | null
  period: string | null
  next_due_at: string | null
  external_id: string | null
  specs: string | null
}

export interface FinanceAccount {
  id: number
  provider_id: number
  provider_name: string
  provider_url: string | null
  adapter: string
  base_url: string | null
  auto_sync: boolean
  low_balance_threshold: number | null
  balance: number | null
  balance_currency: string | null
  services: FinanceService[] | null
  last_sync_at: string | null
  last_sync_status: 'ok' | 'error' | null
  last_sync_error: string | null
}

export interface BalanceSnapshot {
  account_id: number
  snapshot_date: string
  balance: number
  currency: string
  provider_name: string
}

export interface AccountTestResult {
  status: 'ok' | 'error'
  error?: string
  balance?: number | null
  currency?: string | null
  services?: FinanceService[]
}

export const financeApi = {
  getSummary: async (months = 6): Promise<FinanceSummary> =>
    (await client.get('/finance/summary', { params: { months } })).data,

  getUpcoming: async (days = 30): Promise<{ items: FinanceItem[] }> =>
    (await client.get('/finance/upcoming', { params: { days } })).data,

  listItems: async (params: Record<string, string | number | undefined> = {}): Promise<{ items: FinanceItem[] }> =>
    (await client.get('/finance/items', { params })).data,

  createItem: async (data: ItemPayload): Promise<FinanceItem> =>
    (await client.post('/finance/items', data)).data,

  updateItem: async (id: number, data: Partial<ItemPayload>): Promise<FinanceItem> =>
    (await client.patch(`/finance/items/${id}`, data)).data,

  deleteItem: async (id: number): Promise<void> => {
    await client.delete(`/finance/items/${id}`)
  },

  markPaid: async (id: number, data: { amount?: number; paid_at?: string; comment?: string } = {}): Promise<FinanceItem> =>
    (await client.post(`/finance/items/${id}/paid`, data)).data,

  skipCycle: async (id: number): Promise<FinanceItem> =>
    (await client.post(`/finance/items/${id}/skip`)).data,

  listPayments: async (params: Record<string, string | number | undefined> = {}): Promise<{ items: FinancePayment[] }> =>
    (await client.get('/finance/payments', { params })).data,

  createPayment: async (data: {
    item_id?: number | null; item_name?: string; kind: string
    paid_at: string; amount: number; currency: string; comment?: string
  }): Promise<{ id: number }> =>
    (await client.post('/finance/payments', data)).data,

  deletePayment: async (id: number): Promise<void> => {
    await client.delete(`/finance/payments/${id}`)
  },

  listCategories: async (): Promise<{ items: FinanceCategory[] }> =>
    (await client.get('/finance/categories')).data,

  createCategory: async (data: { name: string; kind: string; color?: string; icon?: string }): Promise<FinanceCategory> =>
    (await client.post('/finance/categories', data)).data,

  deleteCategory: async (id: number): Promise<void> => {
    await client.delete(`/finance/categories/${id}`)
  },

  listProviders: async (): Promise<{ items: FinanceProvider[] }> =>
    (await client.get('/finance/providers')).data,

  createProvider: async (data: { name: string; url?: string; notes?: string }): Promise<FinanceProvider> =>
    (await client.post('/finance/providers', data)).data,

  deleteProvider: async (id: number): Promise<void> => {
    await client.delete(`/finance/providers/${id}`)
  },

  listRates: async (): Promise<{ items: FinanceRate[]; base_currency: string }> =>
    (await client.get('/finance/rates')).data,

  setRate: async (currency: string, rate_rub: number): Promise<void> => {
    await client.put(`/finance/rates/${currency}`, { rate_rub, is_manual: true })
  },

  refreshRates: async (): Promise<{ updated: number; items: FinanceRate[] }> =>
    (await client.post('/finance/rates/refresh')).data,

  importFromPanel: async (currency = 'USD'): Promise<{ providers: number; items: number; payments: number; retagged: number; skipped: number; errors: string[] }> =>
    (await client.post('/finance/import-panel', null, { params: { currency } })).data,

  listAdapters: async (): Promise<{ items: HosterAdapter[] }> =>
    (await client.get('/finance/adapters')).data,

  listAccounts: async (): Promise<{ items: FinanceAccount[] }> =>
    (await client.get('/finance/accounts')).data,

  createAccount: async (data: {
    provider_id: number; adapter: string; base_url?: string | null
    credentials: Record<string, string>; auto_sync?: boolean; low_balance_threshold?: number | null
  }): Promise<FinanceAccount> =>
    (await client.post('/finance/accounts', data)).data,

  updateAccount: async (id: number, data: {
    base_url?: string | null; credentials?: Record<string, string>
    auto_sync?: boolean; low_balance_threshold?: number | null
  }): Promise<FinanceAccount> =>
    (await client.patch(`/finance/accounts/${id}`, data)).data,

  deleteAccount: async (id: number): Promise<void> => {
    await client.delete(`/finance/accounts/${id}`)
  },

  testAccount: async (data: {
    account_id?: number; adapter?: string; base_url?: string | null
    credentials?: Record<string, string>
  }): Promise<AccountTestResult> =>
    (await client.post('/finance/accounts/test', data)).data,

  syncAccount: async (id: number): Promise<{ status: string; error?: string; balance?: number; currency?: string; services?: number; due_dates_updated?: number }> =>
    (await client.post(`/finance/accounts/${id}/sync`)).data,

  listSnapshots: async (days = 90): Promise<{ items: BalanceSnapshot[] }> =>
    (await client.get('/finance/snapshots', { params: { days } })).data,

  getBedolagaIncome: async (): Promise<BedolagaIncome> =>
    (await client.get('/finance/bedolaga-income')).data,
}

export interface BedolagaIncome {
  currency: string
  total: { deposit_income: number; subscription_income: number; profit: number }
  today: { deposit_income: number; transactions_count: number }
  by_payment_method: Record<string, number>
}
